File: solutions/haskell/053-sort-colors.hs
{-
  Problem 53: Sort Colors (LeetCode 75)
  Difficulty: Med
  Language: Haskell
-}
sortColors :: [Int] -> [Int]
sortColors nums =
  let counts = foldl (\[r,w,b] n -> case n of
        0 -> [r+1,w,b]; 1 -> [r,w+1,b]; _ -> [r,w,b+1]) [0,0,0] nums
  in concatMap (\(c, n) -> replicate n c) (zip [0,1,2] counts)

File: solutions/typescript/053-sort-colors.ts
/**
 * Problem 53: Sort Colors (LeetCode 75)
 * Difficulty: Med
 * Language: TypeScript
 */
function sortColors(nums: number[]): number[] {
  const counts = nums.reduce(
    (acc, n) => acc.map((c, i) => (i === n ? c + 1 : c)),
    [0, 0, 0]
  );
  return counts.flatMap((count, color) => Array(count).fill(color));
}
